sszgen/main.go

```diff
@@ -932,81 +921,105 @@ func (e *env) parseASTFieldType(name, tags string, expr ast.Expr) (*Value, error
 		}
 
 	case *ast.ArrayType:
-		if isByte(obj.Elt) {
-			if fixedlen := getObjLen(obj); fixedlen != 0 {
-				// array of fixed size
-				return &Value{t: TypeBytes, c: true, s: fixedlen, n: fixedlen}, nil
+		dims, err := extractSSZDimensions(tags)
+		if err != nil {
+			return nil, err
+		}
+
+		collectionExpr := obj
+		outer := &Value{}
+		collection := outer
+		for _, dim := range dims {
+			if dim.IsVector() {
+				collection.t = TypeVector
+				collection.s = uint64(dim.VectorLen())
 			}
-			// []byte
-			if tag, ok := getTags(tags, "ssz"); ok && tag == "bitlist" {
-				// bitlist requires a ssz-max field
-				max, ok := getTagsInt(tags, "ssz-max")
+			if dim.IsList() {
+				collection.t = TypeList
+				collection.m = uint64(dim.ListLen())
```

Comment:
So vectors have `collection.m` set to 0. I guess we don't have to differentiate between something of max capacity 0 and a vector, but it would be nice if we had a better default value for it in the future. Not important to change in this PR

Reply:
That's right. This PR actually takes a stab at simplifying the whole .s/.n/.m logic quite a bit. For example the code is no longer mutating the value of .s/.n during code generation or checking the value of these attributes to infer types or make processing decisions. We're only using it to keep the value of `ssz-max` for places where we need it templated in.